docs/_build/man/bedtools.1
.TH "BEDTOOLS" "1" "July 21, 2012" "2.16.2" "bedtools"
.SH NAME
bedtools \- Bedtools Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.sp
Brief paragraph of the software.
.SH OVERVIEW
.SS 1.1 Background
.sp
The development of BEDTools was motivated by a need for fast, flexible tools with which to compare large sets of genomic
features. Answering fundamental research questions with existing tools was either too slow or required modifications to the
way they reported or computed their results. We were aware of the utilities on the UCSC Genome Browser and Galaxy websites, as
well as the elegant tools available as part of Jim Kent’s monolithic suite of tools (“Kent source”). However, we found that
the web\-based tools were too cumbersome when working with large datasets generated by current sequencing technologies.
Similarly, we found that the Kent source command line tools often required a local installation of the UCSC Genome Browser.
These limitations, combined with the fact that we often wanted an extra option here or there that wasn’t available with
existing tools, led us to develop our own from scratch. The initial version of BEDTools was publicly released in the spring of
2009. The current version has evolved from our research experiences and those of the scientists using the suite over the last
year. The BEDTools suite enables one to answer common questions of genomic data in a fast and reliable manner. The fact that
almost all the utilities accept input from “stdin” allows one to “stream / pipe” several commands together to facilitate more
complicated analyses. Also, the tools allow fine control over how output is reported. The initial version of BEDTools
supported solely 6\-column \fI\%BED\fP files. \fIHowever, we have subsequently added support for sequence alignments in\fP \fI\%BAM\fP
\fIformat, as well as for features in\fP \fI\%GFF\fP , \fI“blocked” BED format, and\fP
\fI\%VCF\fP \fIformat\fP.
The tools are quite fast and typically finish in a matter of a few seconds, even for large datasets. This manual seeks to describe the behavior and
available functionality for each BEDTool. Usage examples are scattered throughout the text, and formal examples are
provided in the last two sections, we hope that this document will give you a sense of the flexibility of
the toolkit and the types of analyses that are possible with BEDTools. If you have further questions, please join the BEDTools
discussion group, visit the Usage Examples on the Google Code site (usage, advanced usage), or take a look at the nascent
“Usage From the Wild” page.
.SS 1.2 Summary of available tools.
.sp
BEDTools support a  wide range of operations for  interrogating and manipulating genomic features. The table below summarizes
the tools available in the suite.
.TS
center;
|l|l|.
_
T{
Utility
T}	T{
Description
T}
_
T{
\fBintersectBed\fP
T}	T{
Returns overlaps between two BED/GFF/VCF files.
T}
_
T{
\fBpairToBed\fP
T}	T{
Returns overlaps between a paired\-end BED file and a regular BED/VCF/GFF file.
T}
_
T{
\fBbamToBed\fP
T}	T{
Converts BAM alignments to BED6, BED12, or BEDPE format.
T}
_
T{
\fBbedToBam\fP
T}	T{
Converts BED/GFF/VCF features to BAM format.
T}
_
T{
\fBbed12ToBed6\fP
T}	T{
Converts "blocked" BED12 features to discrete BED6 features.
T}
_
T{
\fBbedToIgv\fP
T}	T{
Creates IGV batch scripts for taking multiple snapshots from BED/GFF/VCF features.
T}
_
T{
\fBcoverageBed\fP
T}	T{
Summarizes the depth and breadth of coverage of features in one BED versus features (e.g, windows, exons, etc.) defined in another BED/GFF/VCF file.
T}
_
T{
\fBmultiBamCov\fP
T}	T{
Counts sequence coverage for multiple position\-sorted bams at specific loci defined in a BED/GFF/VCF file
T}
_
T{
\fBtagBam\fP
T}	T{
Annotates a BAM file with custom tag fields based on overlaps with BED/GFF/VCF files
T}
_
T{
\fBnuclBed\fP
T}	T{
Profiles the nucleotide content of intervals in a fasta file
T}
_
T{
\fBgenomeCoverageBed\fP
T}	T{
Creates either a histogram, BEDGRAPH, or a "per base" report of genome coverage.
T}
_
T{
\fBunionBedGraphs\fP
T}	T{
Combines multiple BedGraph? files into a single file, allowing coverage/other comparisons between them.
T}
_
T{
\fBannotateBed\fP
T}	T{
Annotates one BED/VCF/GFF file with overlaps from many others.
T}
_
T{
\fBgroupBy\fP
T}	T{
Deprecated. Now in the filo package.
T}
_
T{
\fBoverlap\fP
T}	T{
Returns the number of bases pairs of overlap b/w two features on the same line.
T}
_
T{
\fBpairToPair\fP
T}	T{
Returns overlaps between two paired\-end BED files.
T}
_
T{
\fBclosestBed\fP
T}	T{
Returns the closest feature to each entry in a BED/GFF/VCF file.
T}
_
T{
\fBsubtractBed\fP
T}	T{
Removes the portion of an interval that is overlapped by another feature.
T}
_
T{
\fBwindowBed\fP
T}	T{
Returns overlaps between two BED/VCF/GFF files based on a user\-defined window.
T}
_
T{
\fBmergeBed\fP
T}	T{
Merges overlapping features into a single feature.
T}
_
T{
\fBcomplementBed\fP
T}	T{
Returns all intervals not spanned by the features in a BED/GFF/VCF file.
T}
_
T{
\fBfastaFromBed\fP
T}	T{
Creates FASTA sequences based on intervals in a BED/GFF/VCF file.
T}
_
T{
\fBmaskFastaFromBed\fP
T}	T{
Masks a FASTA file based on BED coordinates.
T}
_
T{
\fBshuffleBed\fP
T}	T{
Randomly permutes the locations of a BED file among a genome.
T}
_
T{
\fBslopBed\fP
T}	T{
Adjusts each BED entry by a requested number of base pairs.
T}
_
T{
\fBflankBed\fP
T}	T{
Creates flanking intervals for each feature in a BED/GFF/VCF file.
T}
_
T{
\fBsortBed\fP
T}	T{
Sorts a BED file by chrom, then start position. Other ways as well.
T}
_
T{
\fBlinksBed\fP
T}	T{
Creates an HTML file of links to the UCSC or a custom browser.
T}
_
.TE
.SS 1.3 Fundamental concepts.
.SS 1.3.1 What are genome features and how are they represented?
.sp
Throughout this manual, we will discuss how to use BEDTools to manipulate, compare and ask questions of genome “features”. Genome features can be functional elements (e.g., genes), genetic polymorphisms (e.g.
SNPs, INDELs, or structural variants), or other annotations that have been discovered or curated by genome sequencing groups or genome browser groups. In addition, genome features can be custom annotations that
an individual lab or researcher defines (e.g., my novel gene or variant).
.sp
The basic characteristics of a genome feature are the chromosome or scaffold on which the feature “resides”, the base pair on which the
feature starts (i.e. the “start”), the base pair on which feature ends (i.e. the “end”), the strand on which the feature exists (i.e. “+” or “\-“), and the name of the feature if one is applicable.
.sp
The two most widely used formats for representing genome features are the BED (Browser Extensible Data) and GFF (General Feature Format) formats. BEDTools was originally written to work exclusively with genome features
described using the BED format, but it has been recently extended to seamlessly work with BED, GFF and VCF files.
.sp
Existing annotations for the genomes of many species can be easily downloaded in BED and GFF
format from the UCSC Genome Browser’s “Table Browser” (\fI\%http://genome.ucsc.edu/cgi-bin/hgTables?command=start\fP) or from the “Bulk Downloads” page (\fI\%http://hgdownload.cse.ucsc.edu/downloads.html\fP). In addition, the
Ensemble Genome Browser contains annotations in GFF/GTF format for many species (\fI\%http://www.ensembl.org/info/data/ftp/index.html\fP)
.SS 1.3.2 Overlapping / intersecting features.
.sp
Two genome features (henceforth referred to as “features”) are said to overlap or intersect if they share at least one base in common.
In the figure below, Feature A intersects/overlaps Feature B, but it does not intersect/overlap Feature C.
.sp
\fBTODO: place figure here\fP
.SS 1.3.3 Comparing features in file “A” and file “B”.
.sp
The previous section briefly introduced a fundamental naming convention used in BEDTools. Specifically, all BEDTools that compare features contained in two distinct files refer to one file as feature set “A” and the other file as feature set “B”. This is mainly in the interest of brevity, but it also has its roots in set theory.
As an example, if one wanted to look for SNPs (file A) that overlap with exons (file B), one would use intersectBed in the following manner:
.sp
.nf
.ft C
intersectBed –a snps.bed –b exons.bed
.ft P
.fi
.sp
There are two exceptions to this rule: 1) When the “A” file is in BAM format, the “\-abam” option must bed used. For example:
.sp
.nf
.ft C
intersectBed –abam alignedReads.bam –b exons.bed
.ft P
.fi
.sp
And 2) For tools where only one input feature file is needed, the “\-i” option is used. For example:
.sp
.nf
.ft C
mergeBed –i repeats.bed
.ft P
.fi
.SS 1.3.4 BED starts are zero\-based and BED ends are one\-based.
.sp
BEDTools users are sometimes confused by the way the start and end of BED features are represented. Specifically, BEDTools uses the UCSC Genome Browser’s internal database convention of making the start position 0\-based and the end position 1\-based: (\fI\%http://genome.ucsc.edu/FAQ/FAQtracks#tracks1\fP)
In other words, BEDTools interprets the “start” column as being 1 basepair higher than what is represented in the file. For example, the following BED feature represents a single base on chromosome 1; namely, the 1st base:
.sp
.nf
.ft C
chr1   0        1    first_base
.ft P
.fi
.sp
Why, you might ask? The advantage of storing features this way is that when computing the length of a feature, one must simply subtract the start from the end. Were the start position 1\-based,
the calculation would be (slightly) more complex (i.e. (end\-start)+1). Thus, storing BED features this way reduces the computational burden.
.SS 1.3.5 GFF starts and ends are one\-based.
.sp
In contrast, the GFF format uses 1\-based coordinates for both the start and the end positions. BEDTools is aware of this and adjusts the positions accordingly.
In other words, you don’t need to subtract 1 from the start positions of your GFF features for them to work correctly with BEDTools.
.SS 1.3.6 VCF coordinates are one\-based.
.sp
The VCF format uses 1\-based coordinates. As in GFF, BEDTools is aware of this and adjusts the positions accordingly.
In other words, you don’t need to subtract 1 from the start positions of your VCF features for them to work correctly with BEDTools.
.SS 1.3.7 File B is loaded into memory (most of the time).
.sp
Whenever a BEDTool compares two files of features, the “B” file is loaded into memory. By contrast, the “A” file is processed line by line and compared with the features from B.
Therefore to minimize memory usage, one should set the smaller of the two files as the B file. One salient example is the comparison of aligned sequence reads from a
current DNA sequencer to gene annotations.      In this case, the aligned sequence file (in BED format) may have tens of millions of features (the sequence alignments),
while the gene annotation file will have tens of thousands of features. In this case, it is wise to sets the reads as file A and the genes as file B.
.SS 1.3.8 Feature files \fImust\fP be tab\-delimited.
.sp
This is rather self\-explanatory. While it is possible to allow BED files to be space\-delimited, we have decided to require tab delimiters for three reasons:
.INDENT 0.0
.IP 1. 3
By requiring one delimiter type, the processing time is minimized.
.IP 2. 3
Tab\-delimited files are more amenable to other UNIX utilities.
.IP 3. 3
GFF files can contain spaces within attribute columns. This complicates the use of space\-delimited files as spaces must therefore be treated specially depending on the context.
.UNINDENT
.SS 1.3.9 All BEDTools allow features to be “piped” via standard input.
.sp
In an effort to allow one to combine multiple BEDTools and other UNIX utilities into more complicated “pipelines”, all BEDTools allow features
to be passed to them via standard input. Only one feature file may be passed to a BEDTool via standard input.
The convention used by all BEDTools is to set either file A or file B to “stdin” or "\-". For example:
.sp
.nf
.ft C
cat snps.bed | intersectBed –a stdin –b exons.bed
cat snps.bed | intersectBed –a \- –b exons.bed
.ft P
.fi
.sp
In addition, all BEDTools that simply require one main input file (the \-i file) will assume that input is
coming from standard input if the \-i parameter is ignored. For example, the following are equivalent:
.sp
.nf
.ft C
cat snps.bed | sortBed –i stdin
cat snps.bed | sortBed
.ft P
.fi
.SS 1.3.10 Most BEDTools write their results to standard output.
.sp
To allow one to combine multiple BEDTools and other UNIX utilities into more complicated “pipelines”,
most BEDTools report their output to standard output, rather than to a named file. If one wants to write the output to a named file, one can use the UNIX “file redirection” symbol “>” to do so.
Writing to standard output (the default):
.sp
.nf
.ft C
intersectBed –a snps.bed –b exons.bed
chr1 100100 100101 rs233454
chr1 200100 200101 rs446788
chr1 300100 300101 rs645678
.ft P
.fi
.sp
Writing to a file:
.sp
.nf
.ft C
intersectBed –a snps.bed –b exons.bed > snps.in.exons.bed

cat snps.in.exons.bed
chr1 100100 100101 rs233454
chr1 200100 200101 rs446788
chr1 300100 300101 rs645678
.ft P
.fi
.SS 1.3.11 What is a “genome” file?
.sp
Some of the BEDTools (e.g., genomeCoverageBed, complementBed, slopBed) need to know the size of
the chromosomes for the organism for which your BED files are based. When using the UCSC Genome
Browser, Ensemble, or Galaxy, you typically indicate which species / genome build you are working.
The way you do this for BEDTools is to create a “genome” file, which simply lists the names of the
chromosomes (or scaffolds, etc.) and their size (in basepairs).
Genome files must be tab\-delimited and are structured as follows (this is an example for C. elegans):
.sp
.nf
.ft C
chrI 15072421
chrII 15279323
\&...
chrX 17718854
chrM 13794
.ft P
.fi
.sp
BEDTools includes predefined genome files for human and mouse in the /genomes directory included
in the BEDTools distribution. Additionally, the “chromInfo” files/tables available from the UCSC
Genome Browser website are acceptable. For example, one can download the hg19 chromInfo file here:
\fI\%http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/chromInfo.txt.gz\fP
.SS 1.3.12 Paired\-end BED files (BEDPE files).
.sp
We have defined a new file format (BEDPE) to concisely describe disjoint genome features, such as
structural variations or paired\-end sequence alignments. We chose to define a new format because the
existing BED block format (i.e. BED12) does not allow inter\-chromosomal feature definitions. Moreover,
the BED12 format feels rather bloated when one want to describe events with only two blocks.
.SS 1.3.13 Use “\-h” for help with any BEDTool.
.sp
Rather straightforward. If you use the “\-h” option with any BEDTool, a full menu of example usage
and available options (when applicable) will be reported.
.SS 1.3.14 BED features must not contain negative positions.
.sp
BEDTools will typically reject BED features that contain negative positions. In special cases, however,
BEDPE positions may be set to \-1 to indicate that one or more ends of a BEDPE feature is unaligned.
.SS 1.3.15 The start position must be <= to the end position.
.sp
BEDTools will reject BED features where the start position is greater than the end position.
.SS 1.3.16 Headers are allowed in GFF and BED files
.sp
BEDTools will ignore headers at the beginning of BED and GFF files. Valid header lines begin with a
“#” symbol, the work “track”, or the word “browser”. For example, the following examples are valid
headers for BED or GFF files:
.sp
.nf
.ft C
track name=aligned_read description="Illumina aligned reads”
chr5 100000 500000 read1 50 +
chr5 2380000 2386000 read2 60 \-

#This is a fascinating dataset
chr5 100000 500000 read1 50 +
chr5 2380000 2386000 read2 60 \-

browser position chr22:1\-20000
chr5 100000 500000 read1 50 +
chr5 2380000 2386000 read2 60 \-
.ft P
.fi
.SS 1.3.17 GZIP support: BED, GFF, VCF, and BEDPE file can be “gzipped”
.sp
BEDTools will process gzipped BED, GFF, VCF and BEDPE files in the same manner as
uncompressed files. Gzipped files are auto\-detected thanks to a helpful contribution from Gordon
Assaf.
.SS 1.3.18 Support for “split” or “spliced” BAM alignments and “blocked” BED features
.sp
As of Version 2.8.0, five BEDTools (\fBintersectBed\fP, \fBcoverageBed\fP, \fBgenomeCoverageBed\fP,
\fBbamToBed\fP, and \fBbed12ToBed6\fP) can properly handle “split”/”spliced” BAM alignments (i.e., having an
“N” CIGAR operation) and/or “blocked” BED (aka BED12) features.
.sp
\fBintersectBed\fP, \fBcoverageBed\fP, and \fBgenomeCoverageBed\fP will optionally handle “split” BAM and/or
“blocked” BED by using the \fB\-split\fP option. This will cause intersects or coverage to be computed only
for the alignment or feature blocks. In contrast, without this option, the intersects/coverage would be
computed for the entire “span” of the alignment or feature, regardless of the size of the gaps between
each alignment or feature block. For example, imagine you have a RNA\-seq read that originates from
the junction of two exons that were spliced together in a mRNA. In the genome, these two exons
happen to be 30Kb apart. Thus, when the read is aligned to the reference genome, one portion of the
read will align to the first exon, while another portion of the read will align ca. 30Kb downstream to the
other exon. The corresponding CIGAR string would be something like (assuming a 76bp read):
30M*3000N*46M. In the genome, this alignment “spans” 3076 bp, yet the nucleotides in the sequencing
read only align “cover” 76bp. Without the \fB\-split\fP option, coverage or overlaps would be reported for the
entire 3076bp span of the alignment. However, with the \fB\-split\fP option, coverage or overlaps will only
be reported for the portions of the read that overlap the exons (i.e. 30bp on one exon, and
46bp on the other).
.sp
Using the \-split option with bamToBed causes “spliced/split” alignments to be reported in BED12
format. Using the \-split option with bed12ToBed6 causes “blocked” BED12 features to be reported in
BED6 format.
.SS 1.3.19 Writing uncompressed BAM output.
.sp
When working with a large BAM file using a complex set of tools in a pipe/stream, it is advantageous
to pass uncompressed BAM output to each downstream program. This minimizes the amount of time
spent compressing and decompressing output from one program to the next. All BEDTools that create
BAM output (e.g. \fBintersectBed\fP, \fBwindowBed\fP) will now optionally create uncompressed BAM output
using the \fB\-ubam\fP option.
.SS 1.4 Implementation and algorithmic notes.
.sp
BEDTools was implemented in C++ and makes extensive use of data structures and fundamental
algorithms from the Standard Template Library (STL). Many of the core algorithms are based upon the
genome binning algorithm described in the original UCSC Genome Browser paper (Kent et al, 2002).
The tools have been designed to inherit core data structures from central source files, thus allowing
rapid tool development and deployment of improvements and corrections. Support for BAM files is
made possible through Derek Barnett’s elegant C++ API called BamTools.
.SS 1.5 License and availability.
.sp
BEDTools is freely available under a GNU Public License (Version 2) at:
\fI\%http://bedtools.googlecode.com\fP
.SS 1.6 Mailing list.
.sp
A discussion group for reporting bugs, asking questions of the developer and of the user community, as
well as for requesting new features is available at:
\fI\%http://groups.google.com/group/bedtools-discuss\fP
.SS 1.7 Contributors.
.sp
As open\-source software, BEDTools greatly benefits from contributions made by other developers and
users of the tools. We encourage and welcome suggestions, contributions and complaints. This is how
software matures, improves and stays on top of the needs of its user community. The Google Code
(GC) site maintains a list of individuals who have contributed either source code or useful ideas for
improving the tools. In the near future, we hope to maintain a source repository on the GC site in
order to facilitate further contributions. We are currently unable to do so because we use Git for
version control, which is not yet supported by GC.
.SH INSTALLATION
.sp
BEDTools is intended to run in a "command line" environment on UNIX, LINUX and Apple OS X
operating systems. Installing BEDTools involves downloading the latest source code archive followed by
compiling the source code into binaries on your local system. The following commands will install
BEDTools in a local directory on a NIX or OS X machine. Note that the \fB"<version>"\fP refers to the
latest posted version number on \fI\%http://bedtools.googlecode.com/\fP.
.sp
Note: \fIThe BEDTools "makefiles" use the GCC compiler. One should edit the Makefiles accordingly if
one wants to use a different compiler.\fP:
.sp
.nf
.ft C
curl http://bedtools.googlecode.com/files/BEDTools.<version>.tar.gz > BEDTools.tar.gz
tar \-zxvf BEDTools.tar.gz
cd BEDTools\-<version>
make clean
make all
ls bin
.ft P
.fi
.sp
At this point, one should copy the binaries in BEDTools/bin/ to either usr/local/bin/ or some
other repository for commonly used UNIX tools in your environment. You will typically require
administrator (e.g. "root" or "sudo") privileges to copy to usr/local/bin/. If in doubt, contact you
system administrator for help.
.SH QUICK START
.SS Install BEDTools
.sp
.nf
.ft C
curl http://bedtools.googlecode.com/files/BEDTools.<version>.tar.gz > BEDTools.tar.gz
tar \-zxvf BEDTools.tar.gz
cd BEDTools
make clean
make all
sudo cp bin/* /usr/local/bin/
.ft P
.fi
.SS Use BEDTools
.sp
Below are examples of typical BEDTools usage. \fBAdditional usage examples are described in
section 6 of this manual.\fP Using the "\-h" option with any BEDTools will report a list of all command
line options.
.sp
A. Report the base\-pair overlap between the features in two BED files.
.sp
.nf
.ft C
intersectBed \-a reads.bed \-b genes.bed
.ft P
.fi
.sp

src/bedtools.cpp
/*****************************************************************************
  bedtools.cpp

  bedtools command line interface.  
  Thanks to Heng Li, as this interface is inspired and 
  based upon his samtools interface.

  (c) 2009-2011 - Aaron Quinlan
  Quinlan Laboratory
  Department of Public Health Sciences
  Center for Public Health genomics
  University of Virginia
  aaronquinlan@gmail.com
  
  Licenced under the GNU General Public License 2.0 license.
******************************************************************************/
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <string>
#include "version.h"

using namespace std;

// define our program name
#define PROGRAM_NAME "bedtools"

// colors for the term's menu 
#define RESET "\033[m"
#define GREEN "\033[1;32m"
#define BLUE "\033[1;34m"
#define RED "\033[1;31m"

// define our parameter checking macro
#define PARAMETER_CHECK(param, paramLen, actualLen) (strncmp(argv[i], param, min(actualLen, paramLen))== 0) && (actualLen == paramLen)

int annotate_main(int argc, char* argv[]);//
int bamtobed_main(int argc, char* argv[]);//
int bamtofastq_main(int argc, char* argv[]);//
int bed12tobed6_main(int argc, char* argv[]); //
int bedtobam_main(int argc, char* argv[]);//
int bedtoigv_main(int argc, char* argv[]);//
int bedpetobam_main(int argc, char* argv[]);//
int closest_main(int argc, char* argv[]); //
int cluster_main(int argc, char* argv[]); //
int complement_main(int argc, char* argv[]);//
int coverage_main(int argc, char* argv[]); //
int expand_main(int argc, char* argv[]);//
int fastafrombed_main(int argc, char* argv[]);//
int flank_main(int argc, char* argv[]); //
int genomecoverage_main(int argc, char* argv[]);//
int getoverlap_main(int argc, char* argv[]);//
int groupby_main(int argc, char* argv[]);//
int intersect_main(int argc, char* argv[]); //
int links_main(int argc, char* argv[]);//
int maskfastafrombed_main(int argc, char* argv[]);//
int map_main(int argc, char* argv[]); //
int merge_main(int argc, char* argv[]); //
int multibamcov_main(int argc, char* argv[]);//
int multiintersect_main(int argc, char* argv[]);//
int nuc_main(int argc, char* argv[]);//
int pairtobed_main(int argc, char* argv[]);//
int pairtopair_main(int argc, char* argv[]);//
int random_main(int argc, char* argv[]); //
int shuffle_main(int argc, char* argv[]); //
int slop_main(int argc, char* argv[]); //
int sort_main(int argc, char* argv[]); //
int subtract_main(int argc, char* argv[]); //
int tagbam_main(int argc, char* argv[]);//
int unionbedgraphs_main(int argc, char* argv[]);//
int window_main(int argc, char* argv[]); //
int windowmaker_main(int argc, char* argv[]); //
int bedtools_help(void);
int bedtools_faq(void);


int main(int argc, char *argv[])
{
    // make sure the user at least entered a sub_command
    if (argc < 2) return bedtools_help();

    std::string sub_cmd = argv[1];

    // genome arithmetic tools
    if (sub_cmd == "intersect")        return intersect_main(argc-1, argv+1);
    else if (sub_cmd == "window")      return window_main(argc-1, argv+1);
    else if (sub_cmd == "closest")     return closest_main(argc-1, argv+1);
    else if (sub_cmd == "coverage")    return coverage_main(argc-1, argv+1);
    else if (sub_cmd == "map")         return map_main(argc-1, argv+1);
    else if (sub_cmd == "genomecov")   return genomecoverage_main(argc-1, argv+1);
    else if (sub_cmd == "merge")       return merge_main(argc-1, argv+1);
    else if (sub_cmd == "cluster")     return cluster_main(argc-1, argv+1);    
    else if (sub_cmd == "complement")  return complement_main(argc-1, argv+1);
    else if (sub_cmd == "subtract")    return subtract_main(argc-1, argv+1);
    else if (sub_cmd == "slop")        return slop_main(argc-1, argv+1);
    else if (sub_cmd == "flank")       return flank_main(argc-1, argv+1);
    else if (sub_cmd == "sort")        return sort_main(argc-1, argv+1);
    else if (sub_cmd == "random")      return random_main(argc-1, argv+1);
    else if (sub_cmd == "shuffle")     return shuffle_main(argc-1, argv+1);
    else if (sub_cmd == "annotate")    return annotate_main(argc-1, argv+1);

    // Multi-way file comparisonstools
    else if (sub_cmd == "multiinter")  return multiintersect_main(argc-1, argv+1);
    else if (sub_cmd == "unionbedg")   return unionbedgraphs_main(argc-1, argv+1);

    // paired-end conversion tools
    else if (sub_cmd == "pairtobed")   return pairtobed_main(argc-1, argv+1);
    else if (sub_cmd == "pairtopair")  return pairtopair_main(argc-1, argv+1);

    // format conversion tools
    else if (sub_cmd == "bamtobed")    return bamtobed_main(argc-1, argv+1);
    else if (sub_cmd == "bedtobam")    return bedtobam_main(argc-1, argv+1);
    else if (sub_cmd == "bamtofastq")  return bamtofastq_main(argc-1, argv+1);
    else if (sub_cmd == "bedpetobam")  return bedpetobam_main(argc-1, argv+1);
    else if (sub_cmd == "bed12tobed6") return bed12tobed6_main(argc-1, argv+1);

    // BAM-specific tools
    else if (sub_cmd == "multicov")    return multibamcov_main(argc-1, argv+1);
    else if (sub_cmd == "tag")         return tagbam_main(argc-1, argv+1);

    // fasta tools
    else if (sub_cmd == "getfasta")    return fastafrombed_main(argc-1, argv+1);
    else if (sub_cmd == "maskfasta")   return maskfastafrombed_main(argc-1, argv+1);
    else if (sub_cmd == "nuc")         return nuc_main(argc-1, argv+1);

    // misc. tools
    else if (sub_cmd == "overlap")     return getoverlap_main(argc-1, argv+1);
    else if (sub_cmd == "igv")         return bedtoigv_main(argc-1, argv+1);
    else if (sub_cmd == "links")       return links_main(argc-1, argv+1);
    else if (sub_cmd == "makewindows") return windowmaker_main(argc-1, argv+1);
    else if (sub_cmd == "groupby")     return groupby_main(argc-1, argv+1);
    else if (sub_cmd == "expand")      return expand_main(argc-1, argv+1);

    // help
    else if (sub_cmd == "-h" || sub_cmd == "--help" ||
             sub_cmd == "-help")
        return bedtools_help();

    // frequently asked questions
    else if (sub_cmd == "--FAQ" || sub_cmd == "--faq" ||
             sub_cmd == "-FAQ"  || sub_cmd == "-faq")
        return bedtools_faq();

    // verison information
    else if (sub_cmd == "-version" || sub_cmd == "--version")
        cout << "bedtools " << VERSION << endl;

    // verison information
    else if (sub_cmd == "-contact" || sub_cmd == "--contact")
    {
        cout << endl;
        cout << "- For further help, or to report a bug, please " << endl;
        cout << "  email the bedtools mailing list: " << endl;
        cout << "     bedtools-discuss@googlegroups.com" << endl << endl;

        cout << "- Stable releases of bedtools can be found at: " << endl;
        cout << "     http://bedtools.googlecode.com" << endl << endl;

        cout << "- The development repository can be found at: " << endl;
        cout << "     https://github.com/arq5x/bedtools" << endl << endl;
    }
    // unknown
    else {
        // TODO: Implement a Levenstein-based "did you mean???"
        cerr << "error: unrecognized command: " << argv[1] << endl << endl;
        return 1;
    }
    return 0;
}

int bedtools_help(void)
{
    cout  << PROGRAM_NAME  << ": flexible tools for genome arithmetic and DNA sequence analysis.\n";
    cout << "usage:    bedtools <subcommand> [options]" << endl << endl;

    cout  << "The bedtools sub-commands include:" << endl;
    
    cout  << endl;
    cout  << "[ Genome arithmetic ]" << endl;
    cout  << "    intersect     "  << "Find overlapping intervals in various ways.\n";
    cout  << "    window        "  << "Find overlapping intervals within a window around an interval.\n";
    cout  << "    closest       "  << "Find the closest, potentially non-overlapping interval.\n";    
    cout  << "    coverage      "  << "Compute the coverage over defined intervals.\n";
    cout  << "    map           "  << "Apply a function to a column for each overlapping interval.\n";
    cout  << "    genomecov     "  << "Compute the coverage over an entire genome.\n";
    cout  << "    merge         "  << "Combine overlapping/nearby intervals into a single interval.\n";
    cout  << "    cluster       "  << "Cluster (but don't merge) overlapping/nearby intervals.\n";
    cout  << "    complement    "  << "Extract intervals _not_ represented by an interval file.\n";
    cout  << "    subtract      "  << "Remove intervals based on overlaps b/w two files.\n";
    cout  << "    slop          "  << "Adjust the size of intervals.\n";
    cout  << "    flank         "  << "Create new intervals from the flanks of existing intervals.\n";
    cout  << "    sort          "  << "Order the intervals in a file.\n";
    cout  << "    random        "  << "Generate random intervals in a genome.\n";
    cout  << "    shuffle       "  << "Randomly redistrubute intervals in a genome.\n";
    cout  << "    annotate      "  << "Annotate coverage of features from multiple files.\n";
    
    cout  << endl;
    cout  << "[ Multi-way file comparisons ]" << endl;
    cout  << "    multiinter    "  << "Identifies common intervals among multiple interval files.\n";
    cout  << "    unionbedg     "  << "Combines coverage intervals from multiple BEDGRAPH files.\n";

    cout  << endl;
    cout  << "[ Paired-end manipulation ]" << endl;
    cout  << "    pairtobed     "  << "Find pairs that overlap intervals in various ways.\n";
    cout  << "    pairtopair    "  << "Find pairs that overlap other pairs in various ways.\n";

    cout  << endl;
    cout  << "[ Format conversion ]" << endl;
    cout  << "    bamtobed      "  << "Convert BAM alignments to BED (& other) formats.\n";
    cout  << "    bedtobam      "  << "Convert intervals to BAM records.\n";
    cout  << "    bedtofastq    "  << "Convert BAM records to FASTQ records.\n";
    cout  << "    bedpetobam    "  << "Convert BEDPE intervals to BAM records.\n";    
    cout  << "    bed12tobed6   "  << "Breaks BED12 intervals into discrete BED6 intervals.\n";

    cout  << endl;
    cout  << "[ Fasta manipulation ]" << endl;
    cout  << "    getfasta      "  << "Use intervals to extract sequences from a FASTA file.\n";
    cout  << "    maskfasta     "  << "Use intervals to mask sequences from a FASTA file.\n";
    cout  << "    nuc           "  << "Profile the nucleotide content of intervals in a FASTA file.\n";

    cout  << endl;
    cout  << "[ BAM focused tools ]" << endl;
    cout  << "    multicov      "  << "Counts coverage from multiple BAMs at specific intervals.\n";
    cout  << "    tag           "  << "Tag BAM alignments based on overlaps with interval files.\n";

    cout  << endl;
    cout  << "[ Miscellaneous tools ]" << endl;
    cout  << "    overlap       "  << "Computes the amount of overlap from two intervals.\n"; 
    cout  << "    igv           "  << "Create an IGV snapshot batch script.\n";
    cout  << "    links         "  << "Create a HTML page of links to UCSC locations.\n";
    cout  << "    makewindows   "  << "Make interval \"windows\" across a genome.\n";
    cout  << "    groupby       "  << "Group by common cols. & summarize oth. cols. (~ SQL \"groupBy\")\n";
    cout  << "    expand        "  << "Replicate lines based on lists of values in columns.\n";

    cout  << endl;
    cout  << "[ General help ]" << endl;
    cout  << "    --help        "  << "Print this help menu.\n";
    //cout  << "    --faq         "  << "Frequently asked questions.\n";  TODO
    cout  << "    --version     "  << "What version of bedtools are you using?.\n";
    cout  << "    --contact     "  << "Feature requests, bugs, mailing lists, etc.\n";

    cout << "\n";
    return 0;
}


int bedtools_faq(void)
{
    cout << "\n";

    cout << "Q1. How do I see the help for a given command?" << endl;
    cout << "A1. All BEDTools commands have a \"-h\" option. Additionally, some tools " << endl;
    cout << "    will provide the help menu if you just type the command line " << endl;
    cout << "    followed by enter. " << endl;

    cout << "\n";
    return 0;
}
